README.md
# Document Nav

Document navigation is automatically generated based on the HTML title (h1-h6) tag.

## Install

```
npm i @rsuite/document-nav --save
```

## Examples

```js
import { Content, Nav } from '@rsuite/document-nav';
import '@rsuite/document-nav/lib/less/index.less';

function render() {
  return (
    <div className="row">
      <div className="col col-md-4">
        <Nav />
      </div>
      <div className="col col-md-20">
        <Content>{require('redeme.md')}</Content>
      </div>
    </div>
  );
}
```

## Style

```less
@import '@rsuite/document-nav/lib/less/index.less';
@doc-nav-active-color: #34c3ff;
```

## 使用方法

有**自动解析**和**手动设置**两种使用方式：

### 自动解析

只需按照如下结构组合代码，document-nav 会自动从 `Content` 中解析出所有的 `h` 标签，并生成导航菜单。

```
  <Nav />
  <Content>
    ...content
  </Content>
```

可以使用 `minLevel` 和 `maxLevel` 来限制导航的标题级别，如 `minLevel = 2` 且 `maxLevel = 4` 时，只有 `h2, h3, h4` 会被导航。

`once` 设置为文档只解析一次，用于文章结构不会改变的情况，避免比较解析造成的性能损耗。

`deep` 设置为最大解析深度，通常标题所在的层级不会太深，如果文档 DOM 很复杂同时解析层级过深的话会大大影响性能。

**注意**

该方法使用标题作为锚点 ID，请尽量避免标题名称重复。

```js
import { Content, Nav } from '@rsuite/document-nav';
import '@rsuite/document-nav/lib/less/index.less';

function render() {
  return (
    <div className="row">
      <div className="col col-md-4">
        <Nav />
      </div>
      <div className="col col-md-20">
        <Content>{require('redeme.md')}</Content>
      </div>
    </div>
  );
}
```

### 手动设置

在 Nav 组件中通过 NavItem 组件设置导航。

```js
import { Content, Nav } from '@rsuite/document-nav';
import '@rsuite/document-nav/lib/less/index.less';

function render() {
  return (
    <div className="row">
      <div className="col col-md-4">
        <Nav>
          <Nav.Item anchor="h-3" title="二级标题---2">
              <Nav.Item anchor="h-3-1" title="三级标题">
            </Nav.Item>
            <Nav.Item anchor="h-4" title="二级标题---3" />
        </Nav>
      </div>
      <div className="col col-md-20">
        <Content>{require('redeme.md')}</Content>
      </div>
    </div>
  );
}
```

## Props

### `<Nav>`

| Name            | Type `(Default)`                   | Description                                                                                            |
| --------------- | ---------------------------------- | ------------------------------------------------------------------------------------------------------ |
| fixed           | boolean `(true)`                   | 是否固定在屏幕的某一个位置？仅当为 `true` 时，`offset` 属性才会生效，为 `false` 时，导航菜单跟随文档流 |
| offset          | object `({left: 'auto', top: 60})` | 定位，属性为 `top`, `right`, `bottom`, `left`                                                          |
| width           | number `(250)`                     | 导航部分的宽。当 `fixed` 为 `true` 时，默认为 `250`， 为 `false` 时默认为 `100%`                       |
| scrollBar       | string `('right')`                 | 导航栏边上的滚动条位置，`left` 或 `right`                                                              |
| minLevel        | number `(2)`                       | 最小导航级别，默认为 `2` 即 `h2`                                                                       |
| maxLevel        | number `(4)`                       | 最大导航级别，默认为 `4` 即 `h4`                                                                       |
| showOrderNumber | boolean `(true)`                   | 是否显示标题前的序号                                                                                   |
| once            | boolean `(true)`                   | 只解析一次文档？设为`true`则只会解析一次，文档发生变化时不会重新解析，性能较好。                       |
| rtl             | boolean `(false)`                  | 文档从右侧开始阅读                                                                                     |
| deep            | number `(10)`                      | 解析的 DOM 深度，避免和标题无关的 DOM 层级过深导致的性能损耗                                           |

### `<Nav.Item>`

| Name     | Type `(Default)` | Description      |
| -------- | ---------------- | ---------------- |
| anchor   | string           | 对应的锚点 id    |
| title    | string           | 显示的标题       |
| children | node             | 可选，子导航菜单 |

## License

MIT licensed


CHANGELOG.md
## 1.0.10

修复了在服务端渲染下编译失败及无法正常使用的问题

更换了获取文章 DOM 的方式，`NavProvider` 组件暂时不需要了，未来可能会弃用

## 1.0.9

- 使用 `MutationObserver` API 监听 DOM 树修改，实时解析更新 Nav
- 增加 `deep` 属性，配置最大解析的 DOM 深度，避免解析层级过深，造成卡顿

_考虑在未来性能优化到比较合适后，去掉 `once` 属性_

## 1.0.8

修复了监听事件未正常更新引起的一些问题

## 1.0.6

增加了 rtl 配置


src/Nav.js
// @flow
import * as React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import createNavItems from './utils/createNavItems';
import NavItem from './NavItem';
import { itemHeight } from './constants';
import throttle from './utils/throttle';
import NavContext, { NavItemContext } from './NavContext';

type Props = {
  offset: {
    top?: number | 'auto',
    right?: number | 'auto',
    bottom?: number | 'auto',
    left?: number | 'auto'
  },
  minLevel: number,
  maxLevel: number,
  width?: number,
  scrollBar: 'left' | 'right',
  fixed: boolean | 'vertical',
  showOrderNumber: boolean,
  once: boolean,
  children?: React.Element<typeof NavItem>,
  scrollBar: string,
  activeAnchor: string,
  showOrderNumber: boolean,
  rtl: boolean
};

type State = {
  anchors: Array<string>,
  activeAnchor?: string,
  navItems?: React.Node
};

type TitleList = Array<{
  title: string,
  anchor: string,
  level: number
}>;

class Nav extends React.PureComponent<Props, State> {
  static defaultProps = {
    offset: {
      top: 60,
      left: 'auto'
    },
    minLevel: 2,
    maxLevel: 4,
    deep: 10,
    scrollBar: 'right',
    show: true,
    fixed: true,
    showOrderNumber: true,
    once: true,
    rtl: false
  };

  static Item = NavItem;

  constructor(props: Props) {
    super(props);
    this.state = {
      anchors: []
    };
    this.reload = throttle(() => {
      if (typeof document === 'undefined') {
        return;
      }
      this.handelContentMount(document.querySelector('.rs-document-nav-content'), this.props.rtl);
    }, 200);
  }
  scrollListener: ?any;
  resizeListener: ?any;
  scrollWrap: ?HTMLElement;
  pageNav: ?HTMLElement;
  prevInnerHTML: ?string; // 兼容没有 MutationObserver 的浏览器

  getContext() {
    const { scrollBar, fixed, showOrderNumber } = this.props;
    const { anchors, activeAnchor = fixed ? anchors[0] : '' } = this.state;
    return {
      scrollBar,
      activeAnchor,
      showOrderNumber
    };
  }

  componentDidMount() {
    this.reload();
  }

  componentDidUpdate(nextProps: Props, nextState: State) {
    const { once, rtl } = this.props;
    if (rtl !== nextProps.rtl) {
      this.reload();
      return;
    }
    if (once && !this.props.content && nextProps.content) {
      this.reload();
      return;
    }
    if (typeof window === 'undefined') {
      return;
    }
    // 无 MutationObserver 时使用传统判断

    if (!once && !window.MutationObserver) {
      // 这里的 content 引用值理论上是不会变的，其实不需要这个判断，暂且先留着吧
      if (this.props.content !== nextProps.content) {
        this.reload();
        return;
      }
      if (nextProps.content && this.prevInnerHTML !== nextProps.content.innerHTML) {
        this.reload();
        return;
      }
    }
  }
  componentWillUnmount() {
    if (typeof window === 'undefined') {
      return;
    }
    window.removeEventListener('scroll', this.scrollListener);
    window.removeEventListener('resize', this.resizeListener);
    this.pageNav = null;
  }

  setScrollListener(ref: HTMLElement, anchors: Array<string>) {
    if (typeof window === 'undefined') {
      return;
    }
    this.scrollWrap = ref;
    const elList = anchors.map(anchor => document.getElementById(anchor));
    if (this.scrollListener) {
      window.removeEventListener('scroll', this.scrollListener);
    }
    this.scrollListener = throttle(() => {
      let index = 0;
      const { activeAnchor } = this.state;
      elList.find((el, i) => {
        if (!el) {
          return false;
        }
        const position = el.getBoundingClientRect();
        index = i;
        return position.top > 0;
      });
      const nextAnchor = anchors[index - 1] || anchors[0];
      if (nextAnchor !== activeAnchor && this.pageNav) {
        this.setState({
          activeAnchor: nextAnchor
        });
        const nav = document.querySelector(`a[href='#${nextAnchor}']`);
        const pageNav = this.pageNav;
        if (nav && pageNav) {
          const navTop = nav.getBoundingClientRect().top - pageNav.getBoundingClientRect().top;
          const maxCount = parseInt(pageNav.clientHeight / nav.offsetHeight, 10);
          if (navTop + itemHeight * 2 > pageNav.clientHeight) {
            pageNav.scrollTop = pageNav.scrollTop + (maxCount - 1) * itemHeight;
          }
          if (navTop < 2 * itemHeight) {
            pageNav.scrollTop = pageNav.scrollTop - (maxCount - 1) * itemHeight;
          }
        }
      }
    }, 300);
    window.addEventListener('scroll', this.scrollListener);
  }

  // 遍历所有标题
  traverseTitle(node: HTMLElement, titleList: TitleList, anchors: Array<string>, deep) {
    if (!node || !titleList || !anchors || deep > this.props.deep) {
      return false;
    }
    const type = node.tagName;
    if (/^H[1-6]/.test(type)) {
      const title = node.innerText || '';
      node.id = title;

      let level = parseInt(type.replace('H', ''), 10);
      const { minLevel, maxLevel } = this.props;
      if (level >= minLevel && level <= maxLevel) {
        titleList.push({
          title,
          anchor: title,
          level
        });
        anchors.push(title);
      }
    } else {
      const { children } = node;
      for (let i = 0; i < children.length; i += 1) {
        this.traverseTitle(children[i], titleList, anchors, deep + 1);
      }
    }
  }

  getNavItems() {
    let anchors = [];
    const { children, scrollBar = 'left', rtl } = this.props;
    const { activeAnchor } = this.state;
    const navItems = React.Children.map(children, (item, i) => {
      anchors.push(item.props.anchor);
      return React.cloneElement(item, {
        index: `${i + 1}`,
        level: 1,
        activeAnchor,
        scrollBar,
        key: `${i + 1} ${item.props.anchor}`,
        rtl
      });
    });
    this.setState({
      anchors
    });

    return navItems;
  }

  handelContentMount(content: HTMLElement, rtl) {
    if (typeof window === 'undefined') {
      return;
    }
    if (content) {
      if (window.MutationObserver) {
        this.observe = new MutationObserver(() => {
          if (!this.props.once) {
            this.reload();
          }
        });
        const config = {
          childList: true,
          subtree: true
        };
        this.observe.observe(content, config);
      } else {
        this.prevInnerHTML = content.innerHTML;
      }
    }
    const titleList: TitleList = [];
    const anchors: Array<string> = [];
    const { children, minLevel, maxLevel, fixed } = this.props;
    if (!children) {
      this.traverseTitle(content, titleList, anchors, 0);
      this.setState({
        anchors
      });
      const list = titleList.filter(item => item.level >= minLevel && item.level <= maxLevel);
      const navItems = createNavItems(list, 0, undefined, rtl);
      this.setState({
        navItems
      });
    } else {
      this.setState({
        navItems: this.getNavItems()
      });
    }
    this.setScrollListener(content, anchors);
  }

  bindPageNavRef(nav: React.Ref<any>): any {
    if (!nav) {
      return false;
    }
    const { offset } = this.props;

    this.pageNav = nav;

    if (this.resizeListener) {
      window.removeEventListener('resize', this.resizeListener);
    }
    const resizeListener = () => {
      const pageNav = this.pageNav;
      if (pageNav) {
        pageNav.style.height = `${itemHeight *
          parseInt((window.innerHeight - (offset.top || offset.bottom) - 60) / itemHeight, 10)}px`;
      }
    };

    window.addEventListener('resize', resizeListener);
    this.resizeListener = resizeListener;

    resizeListener();
  }

  render() {
    const { offset, width, fixed, style, className } = this.props;
    const { navItems } = this.state;
    const styles = {
      width: fixed ? width || 250 : width || '100%',
      position: fixed ? 'fixed' : 'relative',
      ...(fixed ? offset : {})
    };
    return (
      <NavItemContext.Provider value={this.getContext()}>
        <div
          className={classnames('document-nav', className)}
          style={{ ...styles, ...style }}
          ref={ref => this.bindPageNavRef(ref)}
        >
          {navItems}
        </div>
      </NavItemContext.Provider>
    );
  }
}


export default Nav;
